Support custom AABBs within MultiMesh resources

puchik opened the pull request:
- Supporting custom AABBs on MultiMesh resources themselves allows us to prevent costly runtime AABB recalculations.
- May also help improve CPUParticle performance.

Fixes #79573 

<!--
Please target the `master` branch in priority.
PRs can target `3.x` if the same change was done in `master`, or is not relevant there.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.
You can mention in the description if the change is compatible with `3.x`.

To speed up the contribution process and avoid CI errors, please set up pre-commit hooks locally:
https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html
-->

puchik committed:
Support custom AABB within MultiMesh resources

- Supporting custom AABB on the MultiMesh resource itself allows us to prevent costly runtime AABB recalculations.
- Should also help improve CPU Particle performance.

clayjohn reviewed (commented):
Looks great! I think this also needs to be exposed to CPUParticles so that they can benefit from the change as well.

GPUParticles3D have a [visibility_aabb](https://docs.godotengine.org/en/latest/classes/class_gpuparticles3d.html#class-gpuparticles3d-property-visibility-aabb) property that maps onto ``particles_custom_aabb``. I think we can implement visibility aabb for CPUParticles as well so the API matches (we could also update the "convert to CPUParticles" code to keep the custom_aabb as well

mohsenph69 commented:
That is great, I can not believe you already did something to solve this problem
So fast
My grass system depend on this, As it change multi-mesh buffer in each grass update
![MTerrain_Grass](https://github.com/godotengine/godot/assets/52196206/db626df8-8fa7-4b57-af25-e27cacc7392a)

puchik committed:
Custom ("visiblity") AABB support for CPUParticles

- Improves performance by reducing time spent on AABB generation.
- Also adds an option to generate the AABB manually in the CPUParticles3D dropdown.

puchik commented on servers/rendering/renderer_rd/storage_rd/mesh_storage.cpp:
```diff
@@ -1484,6 +1484,9 @@ void MeshStorage::_multimesh_mark_all_dirty(MultiMesh *multimesh, bool p_data, b
 
 void MeshStorage::_multimesh_re_create_aabb(MultiMesh *multimesh, const float *p_data, int p_instances) {
 	ERR_FAIL_COND(multimesh->mesh.is_null());
+	if (multimesh->custom_aabb != AABB()) {
+		return;
+	}
```
Fixed/added.

puchik commented:
Thanks for the starting point! I added the `visibility_aabb` property to CPUParticles3D which just passes it forward to the MultiMesh. The converter is also updated and I added the same AABB generator. Seems complete.

puchik commented:
Not sure why the godot-cpp test is failing with an insufficient storage error. Is it a problem with the workflow? Does it need to try running again?

akien-mga commented:
Yes, you need to rebase to get workflow changes that fixed that.

puchik commented:
Okay thanks, just rebased and pushed 👍

clayjohn commented:
Just one final little change. We need to ensure that we aren't notifying dependencies that the AABB has changed when a custom AABB is being used:
```
diff --git a/drivers/gles3/storage/mesh_storage.cpp b/drivers/gles3/storage/mesh_storage.cpp
index a276b73c44..9ed1b29225 100644
--- a/drivers/gles3/storage/mesh_storage.cpp
+++ b/drivers/gles3/storage/mesh_storage.cpp
@@ -1961,9 +1961,10 @@ void MeshStorage::multimesh_set_buffer(RID p_multimesh, const Vector<float> &p_b
        } else if (multimesh->mesh.is_valid()) {
                //if we have a mesh set, we need to re-generate the AABB from the new data
                const float *data = p_buffer.ptr();
-
-               _multimesh_re_create_aabb(multimesh, data, multimesh->instances);
-               multimesh->dependency.changed_notify(Dependency::DEPENDENCY_CHANGED_AABB);
+               if (multimesh->custom_aabb != AABB()) {
+                       _multimesh_re_create_aabb(multimesh, data, multimesh->instances);
+                       multimesh->dependency.changed_notify(Dependency::DEPENDENCY_CHANGED_AABB);
+               }
        }
 }
 
@@ -2110,9 +2111,11 @@ void MeshStorage::_update_dirty_multimeshes() {
                        }
 
                        if (multimesh->aabb_dirty && multimesh->mesh.is_valid()) {
-                               _multimesh_re_create_aabb(multimesh, data, visible_instances);
                                multimesh->aabb_dirty = false;
-                               multimesh->dependency.changed_notify(Dependency::DEPENDENCY_CHANGED_AABB);
+                               if (multimesh->custom_aabb != AABB()) {
+                                       _multimesh_re_create_aabb(multimesh, data, visible_instances);
+                                       multimesh->dependency.changed_notify(Dependency::DEPENDENCY_CHANGED_AABB);
+                               }
                        }
                }
 
diff --git a/servers/rendering/renderer_rd/storage_rd/mesh_storage.cpp b/servers/rendering/renderer_rd/storage_rd/mesh_storage.cpp
index 0451a66066..2f348838b0 100644
--- a/servers/rendering/renderer_rd/storage_rd/mesh_storage.cpp
+++ b/servers/rendering/renderer_rd/storage_rd/mesh_storage.cpp
@@ -1962,9 +1962,10 @@ void MeshStorage::multimesh_set_buffer(RID p_multimesh, const Vector<float> &p_b
        } else if (multimesh->mesh.is_valid()) {
                //if we have a mesh set, we need to re-generate the AABB from the new data
                const float *data = p_buffer.ptr();
-
-               _multimesh_re_create_aabb(multimesh, data, multimesh->instances);
-               multimesh->dependency.changed_notify(Dependency::DEPENDENCY_CHANGED_AABB);
+               if (multimesh->custom_aabb != AABB()) {
+                       _multimesh_re_create_aabb(multimesh, data, multimesh->instances);
+                       multimesh->dependency.changed_notify(Dependency::DEPENDENCY_CHANGED_AABB);
+               }
        }
 }
 
@@ -2083,10 +2084,11 @@ void MeshStorage::_update_dirty_multimeshes() {
}
 
                        if (multimesh->aabb_dirty) {
-                               //aabb is dirty..
-                               _multimesh_re_create_aabb(multimesh, data, visible_instances);
                                multimesh->aabb_dirty = false;
-                               multimesh->dependency.changed_notify(Dependency::DEPENDENCY_CHANGED_AABB);
+                               if (multimesh->custom_aabb != AABB()) {
+                                       _multimesh_re_create_aabb(multimesh, data, visible_instances);
+                                       multimesh->dependency.changed_notify(Dependency::DEPENDENCY_CHANGED_AABB);
+                               }
                        }
                }
```

This saves us having to do a bunch of updates in the RenderingServer

puchik commented:
Updated!

clayjohn reviewed (approved):
Looks great! Thanks for responding quickly to my feedback

akien-mga commented:
Thanks!

Bimbam360 commented:
Assume I should comment here, I just tried applying this in 4.3dev4 and get the following odd behaviour:

**Base and when set to (-1000, -1000, -1000, 2000, 2000, 2000) manually in inspector**
![image](https://github.com/godotengine/godot/assets/11803603/d0e40aa6-1843-4f1c-a327-83ea9d35e3b7)


**When set in GDScript with** `MMInst.multimesh.custom_aabb = AABB(Vector3(-1000, -1000, -1000), Vector3(2000, 2000, 2000))` 

![image](https://github.com/godotengine/godot/assets/11803603/7751f543-a96b-40b3-ae62-df0c0e47db1d)

clayjohn commented:
> Assume I should comment here, I just tried applying this in 4.3dev4 and get the following odd behaviour:
> 
> **Base and when set to (-1000, -1000, -1000, 2000, 2000, 2000) manually in inspector** ![image](https://private-user-images.githubusercontent.com/11803603/309115736-d0e40aa6-1843-4f1c-a327-83ea9d35e3b7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDkyNTcwOTgsIm5iZiI6MTcwOTI1Njc5OCwicGF0aCI6Ii8xMTgwMzYwMy8zMDkxMTU3MzYtZDBlNDBhYTYtMTg0My00ZjFjLWEzMjctODNlYTlkMzVlM2I3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzAxVDAxMzMxOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWVhMjE2ZmY4YWYwYjUzMjJkYWE4ZGZmZDgwMjFiMDAwYTQyMjU5ZTA0MjhlNTFjMjdmYjBmOGNhMDAyNDI4MGQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.wMR9a_9VrMvoB-Y3BAN08peKQZc1LaS1dZVJ8_6pSog)
> 
> **When set in GDScript prior to (or after) adding to the scene with** `MMInst.multimesh.custom_aabb = AABB(Vector3(-1000, -1000, -1000), Vector3(2000, 2000, 2000))`
> 
> ![image](https://private-user-images.githubusercontent.com/11803603/309115710-7751f543-a96b-40b3-ae62-df0c0e47db1d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDkyNTcwOTgsIm5iZiI6MTcwOTI1Njc5OCwicGF0aCI6Ii8xMTgwMzYwMy8zMDkxMTU3MTAtNzc1MWY1NDMtYTk2Yi00MGIzLWFlNjItZGYwYzBlNDdkYjFkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzAxVDAxMzMxOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTY4NzFiMjI2Zjc2YjJlODFlZTQ3NDE3YTQ4MTBiNGRlZjA0MzExMDdjOTFjMzA5ZGM1MGZlZGRjNjllMDlmNzYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.SJZCColK_xV4yFa4v5MyXdiLb-Qu4Q4NU119-cMVPJc)

Could you make a new issue with an MRP? That looks super strange and I can't figure out what the issue is just from looking at it. The fact that the cubes are cut in half doesn't make any sense haha

Bimbam360 commented:
> Could you make a new issue with an MRP? That looks super strange and I can't figure out what the issue is just from looking at it. The fact that the cubes are cut in half doesn't make any sense haha

Done, see https://github.com/godotengine/godot/issues/89028

Also note this seems to be specific to mobile rendered with an imported mesh.
